File: packages/runmycampus-webhook-verifier-js/src/verifier.ts
/**
 * HMAC-SHA256 webhook signature verifier.
 *
 * Dual-runtime: detects WebCrypto (`globalThis.crypto.subtle`) first
 * for browsers + Node 19+ + Edge runtimes; falls back to `node:crypto`
 * when running on older Node releases. Zero runtime dependencies.
 *
 * Header contract (set by the RunMyCampus platform on outbound deliveries):
 *
 *   X-RunMyCampus-Signature  →  sha256=<lowercase-hex-digest>
 *   X-RunMyCampus-Timestamp  →  Unix seconds (integer string)
 *   X-RunMyCampus-Event      →  Dotted event class (e.g. "migration.bundle.completed")
 *   X-RunMyCampus-Version    →  Signature format version ("v1" today)
 *
 * SECURITY: Constant-time compare is non-negotiable. In Node we use
 * `crypto.timingSafeEqual`; in the browser we use a manual XOR fold
 * over both encoded strings (WebCrypto exposes no string-input
 * `timingSafeEqual`). DO NOT replace either with `==` / `===`.
 */

import { BadSignatureError, ClockSkewError, MissingHeaderError, UnsupportedAlgorithmError, VerificationError } from "./errors";

/** Public header names the platform writes; do not hard-code in caller code. */
export const SIGNATURE_HEADER = "X-RunMyCampus-Signature";
export const TIMESTAMP_HEADER = "X-RunMyCampus-Timestamp";
export const EVENT_HEADER = "X-RunMyCampus-Event";
export const VERSION_HEADER = "X-RunMyCampus-Version";

/** Signature prefix that identifies the digest algorithm. */
export const SUPPORTED_PREFIX = "sha256=";

/** Default clock-skew tolerance — matches Stripe / GitHub / Twilio. */
export const DEFAULT_TOLERANCE_SECONDS = 300;

export type BytesLike = string | Uint8Array | ArrayBuffer | ArrayBufferView;

/** Coerce a JS value into a Uint8Array. */
function _toUint8(value: BytesLike | null | undefined): Uint8Array {
  if (value === null || value === undefined) {
    return new Uint8Array(0);
  }
  if (value instanceof Uint8Array) {
    return value;
  }
  if (value instanceof ArrayBuffer) {
    return new Uint8Array(value);
  }
  if (ArrayBuffer.isView(value)) {
    return new Uint8Array(
      (value as ArrayBufferView).buffer,
      (value as ArrayBufferView).byteOffset,
      (value as ArrayBufferView).byteLength,
    );
  }
  if (typeof value === "string") {
    return new TextEncoder().encode(value);
  }
  throw new TypeError("verifier: unsupported body / secret value type");
}

interface SubtleLike {
  importKey(
    format: "raw",
    keyData: BufferSource,
    algorithm: { name: "HMAC"; hash: "SHA-256" },
    extractable: false,
    keyUsages: ["sign"],
  ): Promise<CryptoKey>;
  sign(
    algorithm: "HMAC",
    key: CryptoKey,
    data: BufferSource,
  ): Promise<ArrayBuffer>;
}

function _getSubtle(): SubtleLike | null {
  const g: any = globalThis as any;
  if (g && g.crypto && g.crypto.subtle) return g.crypto.subtle as SubtleLike;
  return null;
}

interface NodeCryptoLike {
  createHmac(alg: string, key: Uint8Array | string): { update(d: Uint8Array): any; digest(enc: "hex"): string };
  timingSafeEqual(a: Uint8Array, b: Uint8Array): boolean;
  webcrypto?: { subtle: SubtleLike };
}

function _getNodeCrypto(): NodeCryptoLike | null {
  // Avoid a literal `require("crypto")` call so bundlers targeting the
  // browser can tree-shake this out. We instead probe `globalThis`'s
  // `process` to ensure we're in Node, then dynamic-`require`.
  const g: any = globalThis as any;
  if (
    typeof g.process === "undefined" ||
    !g.process ||
    !g.process.versions ||
    !g.process.versions.node
  ) {
    return null;
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  try {
    // `eval` keeps tools like esbuild from statically resolving this in
    // browser builds — runtime-only.
    // eslint-disable-next-line no-eval
    const req = (0, eval)("require");
    return req("crypto") as NodeCryptoLike;
  } catch {
    return null;
  }
}

/** Compute the HMAC-SHA256 of `body` under `secret` and return hex (lowercase). */
async function _hmacSha256Hex(
  bodyBytes: Uint8Array,
  secretBytes: Uint8Array,
): Promise<string> {
  const subtle = _getSubtle();
  if (subtle) {
    // Cast through `unknown` — TS 5.7+ tightened DOM `BufferSource` to
    // require ArrayBuffer (not ArrayBufferLike, which could be
    // SharedArrayBuffer). At runtime Uint8Array IS a BufferSource.
    const key = await subtle.importKey(
      "raw",
      secretBytes as unknown as ArrayBuffer,
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["sign"],
    );
    const sigBuf = await subtle.sign("HMAC", key, bodyBytes as unknown as ArrayBuffer);
    return _bytesToHex(new Uint8Array(sigBuf));
  }
  // Node-only fallback (Node 16–18 without webcrypto-as-global).
  const node = _getNodeCrypto();
  if (node && node.createHmac) {
    const mac = node.createHmac("sha256", secretBytes);
    mac.update(bodyBytes);
    return mac.digest("hex");
  }
  throw new VerificationError(
    "verifier: no HMAC-SHA256 crypto backend available in this runtime",
  );
}

function _bytesToHex(bytes: Uint8Array): string {
  let hex = "";
  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i].toString(16);
    hex += b.length === 1 ? "0" + b : b;
  }
  return hex;
}

/**
 * Constant-time string equality over ASCII strings (the signature
 * header form). Always touches the full length regardless of mismatch
 * position.
 *
 * Note: length mismatch IS leaked (same posture as Python
 * `hmac.compare_digest` and Node `crypto.timingSafeEqual`).
 */
function _constantTimeEqual(a: string, b: string): boolean {
  // Prefer Node's `timingSafeEqual` when available — it's hardened in C++.
  const node = _getNodeCrypto();
  if (node && node.timingSafeEqual) {
    const aB = new TextEncoder().encode(a);
    const bB = new TextEncoder().encode(b);
    if (aB.length !== bB.length) return false;
    try {
      return node.timingSafeEqual(aB, bB);
    } catch {
      // fall through to manual fold
    }
  }
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/** Normalize a header value to string (or null when absent / unparseable). */
function _coerceHeader(value: string | Uint8Array | string[] | null | undefined): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value === "string") return value.trim();
  if (value instanceof Uint8Array) {
    try {
      return new TextDecoder("utf-8", { fatal: true }).decode(value).trim();
    } catch {
      return null;
    }
  }
  if (Array.isArray(value) && value.length > 0) {
    return _coerceHeader(value[0]);
  }
  return null;
}

/**
 * Return `sha256=<hex>` for the given raw body + secret.
 * Useful for unit tests and replay tooling. Do NOT use the output
 * to compare against a header with `===`; use `verifySignature` so
 * the compare runs in constant time.
 */
export async function computeSignature(
  body: BytesLike,
  secret: BytesLike,
): Promise<string> {
  const bodyBytes = _toUint8(body);
  const secretBytes = _toUint8(secret);
  const hex = await _hmacSha256Hex(bodyBytes, secretBytes);
  return SUPPORTED_PREFIX + hex;
}

export interface VerifyOptions {
  /** Value of the `X-RunMyCampus-Timestamp` header (string seconds). */
  timestampHeader?: string | Uint8Array | string[] | null;
  /** Max accepted clock skew (seconds). Default 300. */
  toleranceSeconds?: number;
  /** Override for the current unix-time (seconds, float). Useful in tests. */
  now?: number;
}

/**
 * Throw on any failure; resolve to `undefined` on success.
 *
 * Use this when you want to log WHY verification failed (without
 * leaking secret material). All thrown errors derive from
 * {@link VerificationError}.
 */
export async function verifySignatureStrict(
  body: BytesLike,
  signatureHeader: string | Uint8Array | string[] | null | undefined,
  secret: BytesLike,
  opts: VerifyOptions = {},
): Promise<void> {
  const headerStr = _coerceHeader(signatureHeader as any);
  if (!headerStr) {
    throw new MissingHeaderError(
      `webhook verifier: missing or empty ${SIGNATURE_HEADER} header`,
    );
  }
  if (!headerStr.startsWith(SUPPORTED_PREFIX)) {
    throw new UnsupportedAlgorithmError(
      "webhook verifier: signature algorithm not supported by this SDK " +
        "version (only 'sha256=' accepted); upgrade the verifier package.",
    );
  }
  const expected = await computeSignature(body, secret);
  if (!_constantTimeEqual(expected, headerStr)) {
    throw new BadSignatureError(
      "webhook verifier: signature did not match HMAC of the body under " +
        "the given secret (body tampered, wrong secret, or body bytes " +
        "are not the canonical bytes the platform signed)",
    );
  }

  // Clock-skew check is opt-in: enforced only when the timestamp header
  // is provided by the caller.
  if (opts.timestampHeader === undefined || opts.timestampHeader === null) {
    return;
  }
  const tsStr = _coerceHeader(opts.timestampHeader);
  if (!tsStr) {
    throw new MissingHeaderError(
      `webhook verifier: ${TIMESTAMP_HEADER} present but empty`,
    );
  }
  const signedAt = Number(tsStr);
  if (!Number.isFinite(signedAt)) {
    throw new MissingHeaderError(
      `webhook verifier: ${TIMESTAMP_HEADER} is not a numeric unix-seconds value`,
    );
  }
  const tolerance = Number(opts.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS);
  const now = Number(opts.now ?? Date.now() / 1000);
  const skew = Math.abs(now - signedAt);
  if (skew > tolerance) {
    throw new ClockSkewError(
      `webhook verifier: signed timestamp is outside the ${tolerance}-second tolerance window`,
      skew,
    );
  }
}

/**
 * Boolean lenient verifier — resolves to `true` iff every check
 * passes; `false` on ANY failure (no exceptions surfaced).
 *
 * Use this in fail-closed middleware that doesn't need to dispatch
 * on the failure mode.
 */
export async function verifySignature(
  body: BytesLike,
  signatureHeader: string | Uint8Array | string[] | null | undefined,
  secret: BytesLike,
  opts: VerifyOptions = {},
): Promise<boolean> {
  try {
    await verifySignatureStrict(body, signatureHeader, secret, opts);
    return true;
  } catch (err) {
    if (err instanceof VerificationError) return false;
    // Re-throw unexpected errors (e.g. WebCrypto/Node-crypto unavailable).
    // Customer middleware should treat those as a 500, not a 401.
    throw err;
  }
}

File: packages/runmycampus-webhook-verifier-py/src/runmycampus_webhook_verifier/verifier.py
"""HMAC-SHA256 webhook signature verifier.

Pure stdlib (``hmac`` + ``hashlib`` + ``time``); no third-party deps.

Header contract
---------------

The platform writes the following headers on every outbound delivery:

* ``X-RunMyCampus-Signature`` — ``sha256=<lowercase-hex-digest>``
* ``X-RunMyCampus-Timestamp`` — Unix seconds as integer string
* ``X-RunMyCampus-Event`` — dotted event class (``migration.bundle.completed``)
* ``X-RunMyCampus-Version`` — signature format version (``v1`` today)

The verifier accepts the raw request body bytes the platform signed
PLUS the headers, computes the expected HMAC under the customer's
secret, and compares in constant time.

Two API styles
--------------

* :func:`verify_signature` — boolean return, swallows all errors.
  Suitable for "fail-closed-on-any-error" middlewares.
* :func:`verify_signature_strict` — raises a typed
  :class:`~runmycampus_webhook_verifier.exceptions.VerificationError`
  subclass so the caller can log *why* (without leaking the secret).
"""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Optional, Union

from .exceptions import (
    BadSignatureError,
    ClockSkewError,
    MissingHeaderError,
    UnsupportedAlgorithmError,
    VerificationError,
)


# Public header names the platform writes; do not hard-code in caller code.
SIGNATURE_HEADER = "X-RunMyCampus-Signature"
TIMESTAMP_HEADER = "X-RunMyCampus-Timestamp"
EVENT_HEADER = "X-RunMyCampus-Event"
VERSION_HEADER = "X-RunMyCampus-Version"

# Signature prefix that identifies the digest algorithm. Currently only
# ``sha256`` is emitted; the prefix is checked so a future ``sha512=``
# rollout fails closed in old verifiers (rather than silently passing
# whatever a malicious sender supplies).
SUPPORTED_PREFIX = "sha256="

# Default clock-skew tolerance — matches Stripe / GitHub / Twilio.
DEFAULT_TOLERANCE_SECONDS = 300


_BytesLike = Union[str, bytes, bytearray, memoryview]


def _coerce_bytes(value: _BytesLike) -> bytes:
    """Normalize ``str | bytes | bytearray | memoryview`` to ``bytes``."""
    if isinstance(value, str):
        return value.encode("utf-8")
    if isinstance(value, (bytearray, memoryview)):
        return bytes(value)
    return value


def compute_signature(
    body: _BytesLike,
    secret: _BytesLike,
) -> str:
    """Return ``sha256=<hex>`` for the given raw body + secret.

    Useful for unit tests, replay tooling, and debugging signature
    mismatches. Do NOT use the output to compare against a header with
    ``==``; use :func:`verify_signature` so the compare runs in constant
    time.
    """
    body_bytes = _coerce_bytes(body)
    secret_bytes = _coerce_bytes(secret)
    digest = hmac.new(secret_bytes, body_bytes, hashlib.sha256).hexdigest()
    return SUPPORTED_PREFIX + digest


def _coerce_header(value) -> Optional[str]:
    """Normalize a header value to ``str`` (or ``None`` if absent / undecodable)."""
    if value is None:
        return None
    if isinstance(value, str):
        return value.strip()
    if isinstance(value, (bytes, bytearray, memoryview)):
        try:
            return bytes(value).decode("utf-8").strip()
        except UnicodeDecodeError:
            return None
    # Tolerate list-of-strings (some frameworks return that for
    # repeated headers). Take the first.
    if isinstance(value, (list, tuple)) and value:
        return _coerce_header(value[0])
    return None


def verify_signature_strict(
    body: _BytesLike,
    signature_header: Union[str, bytes, None],
    secret: _BytesLike,
    *,
    timestamp_header: Union[str, bytes, None] = None,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: Optional[float] = None,
) -> None:
    """Raise on any verification failure; return ``None`` on success.

    Arguments:
        body: The raw request body bytes. MUST be the exact bytes the
            platform signed — do NOT re-serialize the JSON or change
            key ordering before passing here.
        signature_header: The value of the ``X-RunMyCampus-Signature``
            header.
        secret: Shared HMAC secret the platform gave you at subscription
            creation time. Treat as opaque material.
        timestamp_header: Value of ``X-RunMyCampus-Timestamp``. When
            present, the absolute skew vs ``now`` MUST be within
            ``tolerance_seconds`` or :class:`ClockSkewError` is raised.
            When ``None``, the clock-skew check is skipped — callers
            that need replay defense should always pass this.
        tolerance_seconds: Max accepted skew in seconds. Default 300.
        now: Current unix-time in seconds (float). Defaults to
            ``time.time()``. Pass an explicit value in tests.

    Raises:
        MissingHeaderError: signature header missing / empty.
        UnsupportedAlgorithmError: prefix is not ``sha256=``.
        BadSignatureError: digest mismatch (tamper / wrong secret /
            non-canonical body bytes).
        ClockSkewError: timestamp outside tolerance.

    Security notes:
        Constant-time compare uses :func:`hmac.compare_digest`. We
        intentionally do not echo the bad header value or skew amount
        beyond the magnitude — all error messages are safe to log.
    """
    header_str = _coerce_header(signature_header)
    if not header_str:
        raise MissingHeaderError(
            f"webhook verifier: missing or empty {SIGNATURE_HEADER!r} header"
        )
    if not header_str.startswith(SUPPORTED_PREFIX):
        raise UnsupportedAlgorithmError(
            "webhook verifier: signature algorithm not supported by this "
            "SDK version (only 'sha256=' is accepted); upgrade the verifier "
            "package to a release that supports the new prefix."
        )
    expected = compute_signature(body, secret)
    # hmac.compare_digest accepts equal-length bytes; encode both sides
    # so a unicode-encoded header still compares.
    if not hmac.compare_digest(
        expected.encode("ascii"), header_str.encode("ascii")
    ):
        raise BadSignatureError(
            "webhook verifier: signature did not match HMAC of the body "
            "under the given secret (body tampered, wrong secret, or "
            "body bytes are not the canonical bytes the platform signed)"
        )

    # Clock-skew check is opt-in: only enforced when caller supplied the
    # timestamp header. Customers without replay defense set up should
    # pass it — but a missing timestamp does NOT fail verification (the
    # platform may not always send it for legacy subscriptions).
    if timestamp_header is None:
        return None
    ts_str = _coerce_header(timestamp_header)
    if not ts_str:
        # If caller passed something but it decoded to empty, treat
        # exactly like a missing-header — fail closed.
        raise MissingHeaderError(
            f"webhook verifier: {TIMESTAMP_HEADER!r} present but empty"
        )
    try:
        signed_at = float(ts_str)
    except ValueError as exc:
        raise MissingHeaderError(
            f"webhook verifier: {TIMESTAMP_HEADER!r} is not a numeric "
            "unix-seconds value"
        ) from exc
    current = float(now if now is not None else time.time())
    skew = abs(current - signed_at)
    if skew > float(tolerance_seconds):
        raise ClockSkewError(
            f"webhook verifier: signed timestamp is outside the "
            f"{tolerance_seconds}-second tolerance window",
            skew_seconds=skew,
        )
    return None


def verify_signature(
    body: _BytesLike,
    signature_header: Union[str, bytes, None],
    secret: _BytesLike,
    *,
    timestamp_header: Union[str, bytes, None] = None,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: Optional[float] = None,
) -> bool:
    """Return ``True`` iff every check passes; ``False`` on any failure.

    This is the lenient wrapper around :func:`verify_signature_strict` —
    use this in middlewares that want to fail closed without dispatching
    on the failure mode.

    See :func:`verify_signature_strict` for argument documentation.
    """
    try:
        verify_signature_strict(
            body,
            signature_header,
            secret,
            timestamp_header=timestamp_header,
            tolerance_seconds=tolerance_seconds,
            now=now,
        )
        return True
    except VerificationError:
        return False


__all__ = [
    "SIGNATURE_HEADER",
    "TIMESTAMP_HEADER",
    "EVENT_HEADER",
    "VERSION_HEADER",
    "SUPPORTED_PREFIX",
    "DEFAULT_TOLERANCE_SECONDS",
    "compute_signature",
    "verify_signature",
    "verify_signature_strict",
]
